### ISSUE TITLE
Widening type inference inconsistency

### ISSUE THREAD

**Issue by shamwow** [Microsoft/TypeScript#25855]
**TypeScript Version:**  3.1.0-dev.201xxxxx + whatever version the playground uses

**Search Terms:**
literal 
type widening 

**Code**
Snippet 1:
```ts
type Def = {
  'T': 5,
}

function route<S extends keyof Def>(route: S, handler: () => Def[S]) { }

route('T', function () {
    return 5;
});
```

Snippet 2:
```ts
interface ApiDefBase {
  [path: string]: any;
}

interface AppApiDef extends ApiDefBase {
  'test1': {
    GET: {
      response: 200;
    };
  };
  'test2': {
    POST: {
      response: 200
    }
  };
};

function get<Path extends keyof AppApiDef>(
    path: Path,
    handler: (
      input: AppApiDef[Path]['GET']['params'],
    ) => AppApiDef[Path]['GET']['response'],
): void { }

  
get('test1', function() {
  return 200;
});
```

**Expected behaviour:**
Snippet 1 and snippet 2 should have the same result.

**Actual behavior:**
Snippet 1 raises no error but snippet 2 raises an error: `Type 'number' is not assignable to type '200'.`.

Based on [this](https://github.com/Microsoft/TypeScript/pull/10676), I was expecting snippet 1 to raise an error because I expected `function () { return 5;}` to be inferred as `() => number`. But because of snippet 1's output, I expected snippet 2 to raise no error.

Changing `POST` to `GET` in `AppApiDef` resolves the error being raised in snippet 2.

**Playground Link:** 
[link](http://www.typescriptlang.org/play/#src=%2F%2F%20Snippet%201%0D%0A%0D%0Atype%20Def%20%3D%20%7B%0D%0A%20%20'T'%3A%205%2C%0D%0A%7D%0D%0A%0D%0Afunction%20route%3CS%20extends%20keyof%20Def%3E%0D%0A%20%20%20%20(route%3A%20S%2C%20handler%3A%20()%20%3D%3EDef%5BS%5D)%20%7B%20%7D%0D%0A%0D%0Aroute('T'%2C%20function%20()%20%7B%0D%0A%20%20%20%20return%205%3B%0D%0A%7D)%3B%0D%0A%0D%0A%0D%0A%2F%2F%20Snippet%202%0D%0A%0D%0Ainterface%20ApiDefBase%20%7B%0D%0A%20%20%5Bpath%3A%20string%5D%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20AppApiDef%20extends%20ApiDefBase%20%7B%0D%0A%20%20'test1'%3A%20%7B%0D%0A%20%20%20%20GET%3A%20%7B%0D%0A%20%20%20%20%20%20query%3A%20%7B%20count%3A%20number%20%7D%3B%0D%0A%20%20%20%20%20%20response%3A%20200%3B%0D%0A%20%20%20%20%7D%3B%0D%0A%20%20%7D%3B%0D%0A%20%20'test2'%3A%20%7B%0D%0A%20%20%20%20POST%3A%20%7B%0D%0A%20%20%20%20%20%20response%3A%20200%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%3B%0D%0A%7D%3B%0D%0A%0D%0Afunction%20get%3CPath%20extends%20keyof%20AppApiDef%3E(%0D%0A%20%20%20%20path%3A%20Path%2C%0D%0A%20%20%20%20handler%3A%20(%0D%0A%20%20%20%20%20%20input%3A%20AppApiDef%5BPath%5D%5B'GET'%5D%5B'params'%5D%2C%0D%0A%20%20%20%20)%20%3D%3E%20AppApiDef%5BPath%5D%5B'GET'%5D%5B'response'%5D%2C%0D%0A)%3A%20void%20%7B%20%7D%0D%0A%0D%0A%20%20%0D%0Aget('test1'%2C%20function()%20%7B%0D%0A%20%20return%20200%3B%0D%0A%7D)%3B%0D%0A)

**Related Issues:** 
Didn't see any issues but some relevant pull requests:
- https://github.com/Microsoft/TypeScript/pull/10676
- https://github.com/Microsoft/TypeScript/pull/10676

Perhaps it has something to do with the index signature on `ApiDefBase` because rearranging snippet 2 [to remove the index signature resolves the error](http://www.typescriptlang.org/play/#src=interface%20AppApiDef%20%7B%0D%0A%20%20GET%3A%20%7B%0D%0A%20%20%20%20'test1'%3A%20%7B%0D%0A%20%20%20%20%20%20params%3A%20%7B%7D%2C%0D%0A%20%20%20%20%20%20query%3A%20%7B%20count%3A%20number%20%7D%3B%0D%0A%20%20%20%20%20%20response%3A%20200%3B%0D%0A%20%20%20%20%7D%3B%0D%0A%20%20%7D%0D%0A%20%20POST%3A%20%7B%0D%0A%20%20%20%20'test2'%3A%20%7B%0D%0A%20%20%20%20%20%20response%3A%20200%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%3B%0D%0A%7D%3B%0D%0A%0D%0Afunction%20get%3CPath%20extends%20keyof%20AppApiDef%5B'GET'%5D%3E(%0D%0A%20%20%20%20path%3A%20Path%2C%0D%0A%20%20%20%20handler%3A%20(%0D%0A%20%20%20%20%20%20input%3A%20AppApiDef%5B'GET'%5D%5BPath%5D%5B'params'%5D%2C%0D%0A%20%20%20%20)%20%3D%3E%20AppApiDef%5B'GET'%5D%5BPath%5D%5B'response'%5D%2C%0D%0A)%3A%20void%20%7B%20%7D%0D%0A%0D%0A%20%20%0D%0Aget('test1'%2C%20function()%20%7B%0D%0A%20%20return%20200%3B%0D%0A%7D)%3B%0D%0A)

Originally asked this as a [StackOverflow question](https://stackoverflow.com/questions/51452980/how-does-typescript-inferred-object-literal-type-widening-work/51453239#51453239) but this seems like an inconsistency.

**Comment by ahejlsberg** [comment]
This is working as intended but is a bit involved. During type inference for the call to `get` we need to establish the type of the function expression argument. Since the function expression has no explicit return type annotation, we have to infer from the return type. At the point of making this inference we have not yet established a type argument for `Path` (that's what we're trying to infer), so the contextual type for the return expression is based on the _constraint_ of `AppApiDef[Path]['GET']['response']`. Because of the string index signature inherited from `ApiDefBase`, `Path` is constrained to `string`, i.e. it could be _any_ string. That in turn means that `AppApiDef[Path]['GET']['response']` has the constraint `any`. So, we have no meaningful contextual type for the return expression and we therefore widen to `number`.

If you remove the inherited string index signature (and fix the ensuing errors that were shrouded by it), you'll see that it starts working as intended. I think that's your best option. With the `any` string index signature in place you're really not getting much type checking because _anything_ is permitted.
